testing/adios2/engine/bp/TestBPWriteReadLocalVariables.cpp

```diff
@@ -1754,6 +1754,76 @@ TEST_F(BPWriteReadLocalVariables, ADIOS2BPWriteReadLocal1DSubFile)
     }
 }
 
+TEST_F(BPWriteReadLocalVariables, ADIOS2BPWriteReadLocal2DChangeCount)
+{
+
+    const std::string fname("BPWRLocal2DChangeCount_" + engineName + ".bp");
+
+    int mpiRank = 0;
+    int mpiSize = 1;
+#ifdef ADIOS2_HAVE_MPI
+    MPI_Comm_rank(MPI_COMM_WORLD, &mpiRank);
+    MPI_Comm_size(MPI_COMM_WORLD, &mpiSize);
+#endif
+
+    const size_t Nx0 =
+        static_cast<size_t>(std::pow(2 - static_cast<size_t>(mpiRank), 2)) + 1;
+    const size_t Nx1 = 5;
+    // data for block 0 and block 1 per rank
+    std::vector<float> data(Nx0 * Nx1);
+    const float startBlock = static_cast<float>(3 * mpiRank + 1);
+
+    std::iota(data.begin(), data.end(), startBlock);
+
+#ifdef ADIOS2_HAVE_MPI
+    adios2::ADIOS adios(MPI_COMM_WORLD, adios2::DebugON);
+#else
+    adios2::ADIOS adios(true);
+#endif
+    {
+        adios2::IO io = adios.DeclareIO("TestIO");
+        const adios2::Dims shape{};
+        const adios2::Dims start{};
+        const adios2::Dims count{Nx0, Nx1};
+
+        auto var_r32 = io.DefineVariable<float>("r32", shape, start, count);
+
+        adios2::Engine bpWriter = io.Open(fname, adios2::Mode::Write);
+        bpWriter.Put(var_r32, data.data());
+        bpWriter.Close();
+    }
+#ifdef ADIOS2_HAVE_MPI
+    MPI_Barrier(MPI_COMM_WORLD);
+#endif
+    {
+        adios2::IO io = adios.DeclareIO("ReaderIO");
+#ifdef ADIOS2_HAVE_MPI
+        adios2::Engine bpReader =
+            io.Open(fname, adios2::Mode::Read, MPI_COMM_SELF);
+#else
+        adios2::Engine bpReader = io.Open(fname, adios2::Mode::Read);
+#endif
+
+        auto var_r32 = io.InquireVariable<float>("r32");
+        std::map<size_t, std::vector<float>> dataIn;
```

Comment:
Rather than having to store the data in a map and loop through it after you could do the validation inline for a more compact test:
```c++
std::vector<float> dataIn;
for (size_t b = 0; b < static_cast<size_t>(mpiSize); ++b)
{
    dataIn.clear();
    var_r32.SetBlockSelection(b);
    bpReader.Get(var_r32, dataIn);
    bpReader.PerformGets();

    for (size_t i = 0; i < dataIn[b].size(); ++i)
    {
        EXPECT_EQ(dataIn[i], static_cast<float>(3 * b + 1 + i));
    }
}
```

Reply:
@chuckatkins true, my goal is to reproduce the conditions that triggered this bug, several Gets in deferred mode (for different block sizes going up and down) over several pieces of data. Thanks for checking.

Reply:
Got it.  I often forget what the default mode is, sync or deferred, and I keep thinking it's sync, which is not the case.  Given that, an `std::vector<std::vector<float>>` should be more appropriate than an `std::map<size_t, std::vector<float>>` since your "key" to the map is a zero-based consecutive monotonically increasing integer.

Reply:
Yeah, that's also from the conditions. The call to Get is creating a new map entry.

Reply:
It's the call to `dataIn[b]` that's creating the map entry.  My point is just thaty the map is an unnecessary overhead since you don't need arbitrary comparison functions on key objects since your key range is consecutive integers `[0:mpiSize]`, thus using instead `std::vector<std::vector<float>> dataIn(mpiSize);`

Reply:
the call to `dataIn[b]` in `Get, which makes it a perfectly valid use case regardless of ordering. I don't think we should spend time on this, should we?